File: internal/handler/security_group.go
package handler

import (
	"context"

	"github.com/cloudwego/hertz/pkg/app"

	"k_cockpit/internal/api"
	"k_cockpit/internal/auth"
	"k_cockpit/internal/authz"
	"k_cockpit/internal/securitygroup"
)

// SecurityGroup 提供安全组接口（F-4-03 / F-4-04）。
//
// 权限：管理员管预置组与全部租户的组，租户管自己的。归属校验在服务层
// 统一处理（非所有者返回 404 而非 403，避免通过枚举确认 ID 存在）。
type SecurityGroup struct {
	svc *securitygroup.Service
}

// NewSecurityGroup 构造安全组接口。
func NewSecurityGroup(svc *securitygroup.Service) *SecurityGroup {
	return &SecurityGroup{svc: svc}
}

// ListGroups 返回安全组列表（API-110）。
func (h *SecurityGroup) ListGroups(ctx context.Context, c *app.RequestContext) {
	items, err := h.svc.ListGroups(ctx, int64(queryInt(c, "node_id")), authz.ViewerOf(c))
	if err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, map[string]any{"items": items})
}

type createGroupRequest struct {
	NodeID int64  `json:"node_id"`
	Name   string `json:"name"`
	Remark string `json:"remark"`
	// OwnerID 为空且调用者是管理员时，创建的是系统预置组。
	OwnerID *int64 `json:"owner_id"`
}

// CreateGroup 创建安全组（API-111）。
func (h *SecurityGroup) CreateGroup(ctx context.Context, c *app.RequestContext) {
	var req createGroupRequest
	if err := c.Bind(&req); err != nil {
		api.Fail(c, api.InvalidParameter("请求参数不合法"))
		return
	}

	user := auth.CurrentUser(c)
	info := auth.ClientInfoOf(c)

	view, err := h.svc.CreateGroup(ctx, securitygroup.CreateGroupRequest{
		NodeID: req.NodeID, Name: req.Name, Remark: req.Remark, OwnerID: req.OwnerID,
	}, authz.ViewerOf(c), user.Username, info.IP)
	if err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, view)
}

// DeleteGroup 删除安全组（API-112）。
//
// 仍被网口挂载时拒绝：删掉一个正在被使用的组会**静默地**放开一批机器的
// 流量——那是一次方向与用户预期相反的变更，而且不报任何错。
func (h *SecurityGroup) DeleteGroup(ctx context.Context, c *app.RequestContext) {
	id, err := namedPathID(c, "id", "安全组 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}

	user := auth.CurrentUser(c)
	info := auth.ClientInfoOf(c)

	if err := h.svc.DeleteGroup(ctx, id, authz.ViewerOf(c), user.Username, info.IP); err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, map[string]any{"deleted": true})
}

// ListRules 返回组内规则（API-113）。
func (h *SecurityGroup) ListRules(ctx context.Context, c *app.RequestContext) {
	id, err := namedPathID(c, "id", "安全组 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}

	items, err := h.svc.ListRules(ctx, id, authz.ViewerOf(c))
	if err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, map[string]any{"items": items})
}

type ruleRequest struct {
	Direction   string `json:"direction"`
	Protocol    string `json:"protocol"`
	PortStart   *int   `json:"port_start"`
	PortEnd     *int   `json:"port_end"`
	TargetType  string `json:"target_type"`
	TargetValue string `json:"target_value"`
	Priority    int    `json:"priority"`
	Remark      string `json:"remark"`
}

func (r ruleRequest) toService() securitygroup.RuleRequest {
	return securitygroup.RuleRequest{
		Direction: r.Direction, Protocol: r.Protocol,
		PortStart: r.PortStart, PortEnd: r.PortEnd,
		TargetType: r.TargetType, TargetValue: r.TargetValue,
		Priority: r.Priority, Remark: r.Remark,
	}
}

// CreateRule 新增规则（API-114）。
func (h *SecurityGroup) CreateRule(ctx context.Context, c *app.RequestContext) {
	id, err := namedPathID(c, "id", "安全组 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}
	var req ruleRequest
	if err := c.Bind(&req); err != nil {
		api.Fail(c, api.InvalidParameter("请求参数不合法"))
		return
	}

	user := auth.CurrentUser(c)
	info := auth.ClientInfoOf(c)

	view, err := h.svc.CreateRule(ctx, id, req.toService(),
		authz.ViewerOf(c), user.Username, info.IP)
	if err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, view)
}

// UpdateRule 修改规则（API-115）。
func (h *SecurityGroup) UpdateRule(ctx context.Context, c *app.RequestContext) {
	id, err := namedPathID(c, "id", "安全组 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}
	ruleID, err := namedPathID(c, "ruleID", "规则 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}
	var req ruleRequest
	if err := c.Bind(&req); err != nil {
		api.Fail(c, api.InvalidParameter("请求参数不合法"))
		return
	}

	user := auth.CurrentUser(c)
	info := auth.ClientInfoOf(c)

	view, err := h.svc.UpdateRule(ctx, id, ruleID, req.toService(),
		authz.ViewerOf(c), user.Username, info.IP)
	if err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, view)
}

// DeleteRule 删除规则（API-116）。
func (h *SecurityGroup) DeleteRule(ctx context.Context, c *app.RequestContext) {
	id, err := namedPathID(c, "id", "安全组 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}
	ruleID, err := namedPathID(c, "ruleID", "规则 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}

	user := auth.CurrentUser(c)
	info := auth.ClientInfoOf(c)

	if err := h.svc.DeleteRule(ctx, id, ruleID,
		authz.ViewerOf(c), user.Username, info.IP); err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, map[string]any{"deleted": true})
}

// Attach 把安全组挂到网口（API-117）。
func (h *SecurityGroup) Attach(ctx context.Context, c *app.RequestContext) {
	ifaceID, err := namedPathID(c, "interfaceID", "网口 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}
	groupID, err := namedPathID(c, "id", "安全组 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}

	user := auth.CurrentUser(c)
	info := auth.ClientInfoOf(c)

	if err := h.svc.Attach(ctx, ifaceID, groupID,
		authz.ViewerOf(c), user.Username, info.IP); err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, map[string]any{"attached": true})
}

// Detach 解除挂载（API-118）。主组不能从这里解除。
func (h *SecurityGroup) Detach(ctx context.Context, c *app.RequestContext) {
	ifaceID, err := namedPathID(c, "interfaceID", "网口 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}
	groupID, err := namedPathID(c, "id", "安全组 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}

	user := auth.CurrentUser(c)
	info := auth.ClientInfoOf(c)

	if err := h.svc.Detach(ctx, ifaceID, groupID,
		authz.ViewerOf(c), user.Username, info.IP); err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, map[string]any{"detached": true})
}

// Effective 汇总一台虚拟机的生效规则（API-119）。
//
// 这是 F-4-04 的「汇总并预览」：把挂载的多个组的规则合并去重，并标出每条
// 规则来自哪些组——否则用户看到合并结果后不知道该去哪儿改。
func (h *SecurityGroup) Effective(ctx context.Context, c *app.RequestContext) {
	vmID, err := namedPathID(c, "id", "虚拟机 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}

	preview, err := h.svc.Effective(ctx, vmID, authz.ViewerOf(c))
	if err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, preview)
}

// Apply 应用生效规则（API-120）。
//
// **必须带回预览时的版本号**。用户看到预览、判断没问题、然后点确认，这中间
// 可能有几十秒，而别人完全可能改了其中一个组。不校验版本的话，用户批准的
// 是 A，落下去的是 B——这类问题不报错，只会让某天出现一条谁也想不起来
// 什么时候加的放行规则。
func (h *SecurityGroup) Apply(ctx context.Context, c *app.RequestContext) {
	vmID, err := namedPathID(c, "id", "虚拟机 ID")
	if err != nil {
		api.Fail(c, err)
		return
	}

	var req struct {
		Version string `json:"version"`
	}
	if err := c.Bind(&req); err != nil {
		api.Fail(c, api.InvalidParameter("请求参数不合法"))
		return
	}

	user := auth.CurrentUser(c)
	info := auth.ClientInfoOf(c)

	t, err := h.svc.Apply(ctx, vmID, req.Version,
		authz.ViewerOf(c), user.Username, info.IP)
	if err != nil {
		api.Fail(c, err)
		return
	}
	api.OK(c, map[string]any{"task_id": t.ID, "status": t.Status})
}

File: web/src/api/securitygroup.ts
/**
 * 安全组接口（F-4-03 / F-4-04）。
 *
 * 两件界面必须如实表达的事：
 *
 * 1. **组内只有「允许」规则**。叠加生效意味着生效规则是各组的并集，而并集里
 *    没有「拒绝」的位置——A 组拒绝 22、B 组允许 22，合并后通不通取决于谁先算。
 *    默认拒绝由组的整体语义给出：不在任何允许规则里的流量一律不通。
 * 2. **预览与应用之间绑定版本**。用户确认的必须正是他看到的那一套规则。
 */
import { del, get, patch, post } from './client'

export type Direction = 'ingress' | 'egress'
export type Protocol = 'tcp' | 'udp' | 'icmp' | 'all'
export type TargetType = 'cidr' | 'switch' | 'group'

export interface GroupView {
  id: number
  node_id: number
  owner_id?: number
  name: string
  is_default: boolean
  remark?: string
  rule_count: number
  /** 挂载它的网口数。删除前要告诉用户这个数字。 */
  attached_count: number
  created_at: string
}

export interface RuleView {
  id: number
  group_id: number
  direction: Direction
  protocol: Protocol
  port_start?: number
  port_end?: number
  target_type: TargetType
  target_value?: string
  address_family: string
  /** 只影响**展示顺序**，不影响判定。 */
  priority: number
  remark?: string
  created_at: string
}

export interface EffectiveRule {
  direction: Direction
  protocol: Protocol
  port_start?: number
  port_end?: number
  target_type: TargetType
  target_value?: string
  address_family: string
  /** 贡献了这条规则的组名。必看——否则不知道该去哪儿改。 */
  sources: string[]
}

export interface EffectivePreview {
  vm_id: number
  vm_name: string
  /** 这份快照的指纹。应用时必须带回去。 */
  version: string
  groups: string[]
  rules: EffectiveRule[]
  warnings?: string[]
  interface_count: number
}

export interface RuleInput {
  direction: Direction
  protocol: Protocol
  port_start?: number
  port_end?: number
  target_type: TargetType
  target_value?: string
  priority?: number
  remark?: string
}

export const securityGroupApi = {
  list: (nodeID = 0) =>
    get<{ items: GroupView[] }>(
      '/api/v1/security-groups',
      nodeID > 0 ? { node_id: nodeID } : undefined,
    ),

  create: (input: { node_id: number; name: string; remark?: string }) =>
    post<GroupView>('/api/v1/security-groups', input),

  /**
   * 删除安全组。**仍被网口挂载时会被拒绝**——删掉一个正在被使用的组会
   * 静默地放开一批机器的流量，那是一次方向与预期相反的变更。
   */
  remove: (id: number) => del<{ deleted: boolean }>(`/api/v1/security-groups/${id}`),

  listRules: (groupID: number) =>
    get<{ items: RuleView[] }>(`/api/v1/security-groups/${groupID}/rules`),

  createRule: (groupID: number, input: RuleInput) =>
    post<RuleView>(`/api/v1/security-groups/${groupID}/rules`, input),

  updateRule: (groupID: number, ruleID: number, input: RuleInput) =>
    patch<RuleView>(`/api/v1/security-groups/${groupID}/rules/${ruleID}`, input),

  deleteRule: (groupID: number, ruleID: number) =>
    del<{ deleted: boolean }>(`/api/v1/security-groups/${groupID}/rules/${ruleID}`),

  /** 把安全组挂到网口上（附加组）。 */
  attach: (groupID: number, interfaceID: number) =>
    post<{ attached: boolean }>(
      `/api/v1/security-groups/${groupID}/interfaces/${interfaceID}`,
      {},
    ),

  /** 解除挂载。**主组不能从这里解除**——那等于把网口变成「没有安全组」。 */
  detach: (groupID: number, interfaceID: number) =>
    del<{ detached: boolean }>(
      `/api/v1/security-groups/${groupID}/interfaces/${interfaceID}`,
    ),

  /**
   * 汇总一台虚拟机的生效规则（F-4-04）。
   *
   * 多组叠加、去重，并标出每条规则来自哪些组。返回的 `version` 要在应用时
   * 带回去。
   */
  effective: (vmID: number) =>
    get<EffectivePreview>(`/api/v1/vms/${vmID}/security-groups/effective`),

  /**
   * 应用生效规则。`version` 必须与预览时一致。
   *
   * 版本不匹配会被拒绝而**不会**自动用最新规则继续——自动继续等于替用户
   * 批准了他没看过的东西，那正是要防的事。
   */
  apply: (vmID: number, version: string) =>
    post<{ task_id: number; status: string }>(
      `/api/v1/vms/${vmID}/security-groups/apply`,
      { version },
    ),
}

export const DIRECTION_LABEL: Record<Direction, string> = {
  ingress: '入站',
  egress: '出站',
}

export const PROTOCOL_LABEL: Record<Protocol, string> = {
  tcp: 'TCP',
  udp: 'UDP',
  icmp: 'ICMP',
  all: '全部',
}

/**
 * 协议是否使用端口。
 *
 * ICMP 没有端口概念，`all` 覆盖全部协议因而也不能限定端口——填了会得到一条
 * **看起来有限制、实际没有**的规则。界面上直接禁用端口输入框，而不是等后端
 * 拒绝：那样用户会以为是自己填错了格式。
 */
export function protocolUsesPorts(p: Protocol): boolean {
  return p === 'tcp' || p === 'udp'
}

export const TARGET_LABEL: Record<TargetType, string> = {
  cidr: '网段',
  switch: '交换机',
  group: '安全组',
}

/** 把一条规则渲染成可读的一行。 */
export function ruleText(r: {
  protocol: Protocol
  port_start?: number
  port_end?: number
  target_type: TargetType
  target_value?: string
}): string {
  const proto = PROTOCOL_LABEL[r.protocol] ?? r.protocol
  let port = ''
  if (protocolUsesPorts(r.protocol) && r.port_start != null) {
    port =
      r.port_end != null && r.port_end !== r.port_start
        ? `:${r.port_start}-${r.port_end}`
        : `:${r.port_start}`
  }
  const target = r.target_value ? `${TARGET_LABEL[r.target_type]} ${r.target_value}` : '任意来源'
  return `${proto}${port} ← ${target}`
}
